first steps

kantp opened the pull request:
I've put down a design of how the contract keeps track of the off- and on-chain state, and added a method for issuing tokens. Reocation and attestation will follow soon.

kantp committed:
Start reference implementation.

- Set up data structures for token metadata
- Introduce an abstraction for off-chain storage of a `MerkleMap`

  This allows us to do dependency injection, so that we can use an in-memory map
  for testing, and real-world implementations can use the solution that best
  fits their needs.

- Implement the first method of the `SoulboundToken` smart contract: issuing a
  new token.

kantp committed:
Add minimal description in README.md

kantp committed:
Adding some comments

kantp committed:
Actually require that the signature check succeeds.

kantp committed:
Add validation and revocation.

Revocation does not yet take the RevocationPolicy into account.

kantp committed:
Starting to write tests

teddyjfpender committed:
chore: re-format packages

teddyjfpender commented on src/RevocationPolicy.ts:
```diff
@@ -0,0 +1,22 @@
+    Field
+} from 'o1js';
+
+/** When a token is issued, issuer and holder should agree on who can revoke the token */
+class RevocationPolicy extends Struct ({
+    type: Field,
+}) {
+    public static types = {
+        // the issuer has the right to revoke tokens
+        issuerOnly: Field(0),
+        // only the holder of the token can revoke it
+        holderOnly: Field(1),
+        // a token can only be revoked if issuer and holder agree
+        both: Field(2),
+        // tokens are indestructible
+        neither: Field(3),
+    };
+}
+
+export { RevocationPolicy };
```
There's potentially two ways to use the `RevocationPolicy`
- Instead of making this a `Struct` we can make it another `MerkleMap` who's root is committed to the contract at deployment time and use a `ZkProgram` to attest to the `RevocationPolicy` of the SBT. 

- Alternatively, the `SmartContract` could have a state that defines the `RevociationPolicy` (e.g. a hash or field) and a method check if the input `Struct` argument satisfies those constraints.

teddyjfpender commented on src/SoulboundToken.ts:
```diff
@@ -0,0 +1,122 @@
+ * part of the contract state.
+ * 
+*/
+class SoulboundToken
```
This is looking great!

teddyjfpender committed:
fix: exporting of OffchainMerkleMap directory

kantp committed:
Fix the way we handle the off-chain map.

kantp committed:
Remove the deleted OffchainMerkleMap from index.ts.

This should fix the CI build step.

kantp committed:
Factor out functions driving the token contract in the tests.

This makes the tests more succinct, and also allows us to
keep the off-chain map in sync.

kantp committed:
Add tests that issued tokens are verified, while nonexistent are not.

kantp committed:
Add proper error messages

Also, tried to add a test that demonstrates that verifying a nonexistent token
fails. However, the error that's raised within the smart contract does not get
caught by the test, and the test fails (although running the test shows that the
right thing happens).

kantp committed:
Don't sign the same message for issuing and revoking tokens

If we use the plain metadata, the same signed message that was used to issue a
token will be sufficient to revoke it. We introduce a new type,
`SoulboundRequest` that contains the metadata, as well as the request
type (issuing or revoking a token).

kantp committed:
Add a test that revocation works.

kantp committed:
More tests.

kantp committed:
Check the revocationpolicy when revoking a token.

kantp committed:
Emit events for the merkle root.

This allows users to validate that a token has existed at a certain point,
without needing to call the contract again.
